bindings/java/src/main/native/jucx_common_def.cc

```diff
@@ -232,104 +224,89 @@ static inline void jucx_call_callback(jobject callback, jobject jucx_request,
-            }
-            jucx_context_reset(ctx);
-            ucp_request_free(request);
-        }
-        ucs_recursive_spin_unlock(&ctx->lock);
-    } else {
-        jmethodID empty_constructor = env->GetMethodID(jucx_request_cls, "<init>", "()V");
-        jucx_request = env->NewObject(jucx_request_cls, empty_constructor);
-        set_jucx_request_completed(env, jucx_request, NULL, UCS_PTR_RAW_STATUS(request));
-        if (UCS_PTR_IS_ERR(request)) {
-            JNU_ThrowExceptionByStatus(env, UCS_PTR_STATUS(request));
-            if (callback != NULL) {
-                call_on_error(callback, UCS_PTR_STATUS(request));
-            }
-        } else if (callback != NULL) {
-            call_on_success(callback, jucx_request);
-        }
-    }
-    return jucx_request;
+    jobject jucx_request = reinterpret_cast<jobject>(user_data);
+
+    jucx_request_update_sender_tag(env, jucx_request, info->sender_tag);
+    jucx_request_update_recv_length(env, jucx_request, info->length);
+    jucx_request_callback(request, status, user_data);
 }
 
-jobject process_completed_stream_recv(size_t length, jobject callback)
+void stream_recv_callback(void *request, ucs_status_t status, size_t length, void *user_data)
 {
     JNIEnv *env = get_jni_env();
-    jobject jucx_request = env->NewObject(jucx_request_cls, jucx_request_constructor, NULL);
-    env->SetObjectField(jucx_request, native_id_field, NULL);
-    env->SetLongField(jucx_request, recv_size_field, length);
-    jucx_request_update_status(env, jucx_request, UCS_OK);
+    jobject jucx_request = reinterpret_cast<jobject>(user_data);
+    jucx_request_update_recv_length(env, jucx_request, length);
+
+    jucx_request_callback(request, status, user_data);
+}
+
+jobject jucx_request_allocate(JNIEnv *env, const jobject callback, ucp_request_param_t *param)
+{
+    jobject jucx_request = env->NewObject(jucx_request_cls, jucx_request_constructor);
+    param->op_attr_mask = UCP_OP_ATTR_FIELD_USER_DATA | UCP_OP_ATTR_FIELD_CALLBACK;
+    param->user_data = env->NewGlobalRef(jucx_request);
+
     if (callback != NULL) {
-        jucx_call_callback(callback, jucx_request, UCS_OK);
+         env->SetObjectField(jucx_request, request_callback, callback);
     }
+
     return jucx_request;
 }
 
+void process_request(JNIEnv *env, jobject jucx_request, ucs_status_ptr_t status)
+{
+    // If status is error - throw an exception in java.
+    if (UCS_PTR_IS_ERR(status)) {
+        JNU_ThrowExceptionByStatus(env, UCS_PTR_STATUS(status));
+    }
+
+    // Request completed immidiately. Call jucx callback.
+    if (!UCS_PTR_IS_PTR(status)) {
+        set_jucx_request_completed(env, jucx_request, UCS_PTR_RAW_STATUS(status));
+        jobject callback = env->GetObjectField(jucx_request, request_callback);
+        if (callback != NULL) {
+            jucx_call_callback(callback, jucx_request, UCS_PTR_RAW_STATUS(status));
+            env->SetObjectField(jucx_request, request_callback, NULL);
```

Comment:
why needed?

Reply:
There could be 1 callback object for many requests. I set everywhere to ULL, so requests objects could be garbage collected, and eventually callback as well, (to remove reference).

Reply:
ok, maybe add a comment this is done to remove reference